Raven db 13796

maximburyak opened the pull request:
This PR was initially meant to investigate a failing test (RachisTests.SubscriptionsFailover.SetMentorToSubscriptionWithFailover), but finished up resolving a race condition between a request executor's topology update and a command's failover

maximburyak committed:
RavenDB-13796: Fixing races in tests

maximburyak committed:
RavenDB-13796: Fixing RequestExecutor race condition when we update topology during command failover

ayende commented on src/Raven.Client/Http/NodeSelector.cs:
```diff
@@ -100,6 +100,23 @@ public bool OnUpdateTopology(Topology topology, bool forceUpdate = false)
             return UnlikelyEveryoneFaultedChoice(state);
         }
 
+        public (int Index, ServerNode Node, long TopologyEtag) GetPreferredNodeWithTopology()
```
Isn't this duplicating the code in `GetPreferredNode`? Can we merge these?

maximburyak replied:
done

ppekrol commented on src/Raven.Client/Http/RavenCommand.cs:
```diff
@@ -51,6 +51,8 @@ public abstract class RavenCommand<TResult>
         public bool CanCacheAggressively { get; protected set; }
         public string SelectedNodeTag { get; protected set; }
 
+        public long InitialTopologyEtag = -1;                       
```
internal

maximburyak committed:
RavenDB-13796: Preventing code duplication in GetPrefferedNode* functions

maximburyak commented:
*Looking into failed tests

maximburyak committed:
RavenDB-13796: Preventing from sending twice to first failed node

ppekrol commented on src/Raven.Client/Http/RavenCommand.cs:
```diff
@@ -51,6 +51,8 @@ public abstract class RavenCommand<TResult>
         public bool CanCacheAggressively { get; protected set; }
         public string SelectedNodeTag { get; protected set; }
 
+        internal long FailoverTopologyEtag = -2;                       
```
maybe make it nullable?

ppekrol commented on src/Raven.Client/Http/RequestExecutor.cs:
```diff
@@ -647,6 +647,10 @@ protected virtual bool TryLoadFromCache(JsonOperationContext context)
         {
+            if (command.FailoverTopologyEtag == -2)
+            {                
+                command.FailoverTopologyEtag = _nodeSelector?.GetPreferredNodeWithTopology().TopologyEtag ?? -2;
```
isn't this the same as `RequestExecutor.TopologyEtag` property?

ayende replied:
That is not the same, no. Because the topology may change *while* the command is running

ppekrol replied:
The source of it is the same. `NodeSelector.Topology` which is taken from `State` inside the NodeSelector.

maximburyak committed:
RavenDB-13796:
Setting accessibility level of GetPreferredNodeWithTopology to internal
Introducing a constant for the initial topology value of -2: InitialTopologyEtag
Using NodeSelector's topology property directly rather than calling the GetPreferredNodeWithTopology function
